src/benchmarklib/table_builder.hpp

```diff
@@ -0,0 +1,241 @@
+#pragma once
+
+#include <optional>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "boost/hana/assert.hpp"
+#include "boost/hana/for_each.hpp"
+#include "boost/hana/tuple.hpp"
+#include "boost/hana/zip_with.hpp"
+
+#include "resolve_type.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+// helper structs and functions for compile time, the actual code is in namespace opossum
+namespace table_builder {
+// similar to std::optional but has_value is known at compile time, so "if constexpr" can be used
+// boost::hana::optional does not allow moving and reinitializing its value (or I just did not find out how)
+template <typename T, bool _has_value, typename Enable = void>
+class OptionalConstexpr {
+ public:
+  template <typename... Args>
+  explicit OptionalConstexpr(Args&&... args) : _value{std::forward<Args>(args)...} {}
+
+  static_assert(_has_value);
+  static constexpr bool has_value = true;
+
+  T& value() { return _value; }
+
+ private:
+  T _value;
+};
+
+template <typename T, bool _has_value>
+class OptionalConstexpr<T, _has_value, std::enable_if_t<!_has_value>> {
+ public:
+  template <typename... Args>
+  explicit OptionalConstexpr(Args&&...) {}
+
+  static_assert(!_has_value);
+  static constexpr bool has_value = false;
+
+  T& value() {
+    Assert(false, "empty optional has no value");
+    return {};
+  }
+};
+
+template <typename T, typename Enable = void>
+struct is_optional : std::false_type {};
+
+template <typename T>
+struct is_optional<T, std::enable_if_t<std::is_same_v<T, std::optional<typename T::value_type>>>> : std::true_type {};
+
+// is_optional_v<T> <=> T is of type std::optional<?>
+template <typename T>
+constexpr bool is_optional_v = is_optional<T>::value;
+
+template <typename T, typename Enable = void>
+struct GetValueType {
+  using value_type = T;
+};
+
+template <typename T>
+struct GetValueType<T, std::enable_if_t<is_optional_v<T>>> {
+  using value_type = typename T::value_type;
+};
+
+// get the value_type (lets call it V) of std::optional<V> and V using the same syntax - needed for templating
+template <typename T>
+using get_value_type = typename GetValueType<T>::value_type;
+
+// check if std::optional<V> or V is null using the same syntax - needed for templating
+template <typename T>
+constexpr bool is_null(const T& maybe_optional) {
+  if constexpr (is_optional_v<T>) {
+    return !maybe_optional.has_value();
+  } else {
+    return false;
+  }
+}
+
+// get the value of std::optional<V> or V using the same syntax - needed for templating
+template <typename T>
+get_value_type<T>& get_value(T& maybe_optional) {
+  if constexpr (is_optional_v<T>) {
+    return maybe_optional.value();
+  } else {
+    return maybe_optional;
+  }
+}
+
+}  // namespace table_builder
+
+/**
+ * Helper to build a table with a static column layout, specified by constructor args types and names. Keeps a
```

Comment:
"...specified by template parameters"?

Reply:
I'd argue that the template parameters are not explicitly specified but rather deduced from the ctor args.